### PR TITLE
Unify the comments

### PR DESCRIPTION
**Pull request by ZmagoD**
https://biosistemika.atlassian.net/browse/SCI-474

### CODE REVIEW

File: app/assets/javascripts/comments.js [hunk 1]
```diff
@@ -1,151 +1,319 @@
...
-          // TODO
-        }
-      });
-  });
-}
+            scrollCommentOptions($(parent).find('.dropdown-comment'));
+          },
+          error: function(data) {
+            // Display alert
+            alert(data.responseJSON.message);
```
Comment (ZmagoD) [inline app/assets/javascripts/comments.js]: ?

**Comment by mz3944** [comment]
Boy, I got critics for you, man:

Project comments:
- no date is shown
screenshot from 2016-09-27 1

Task comment:
- no date is shown
- duplicates - circa 20 first old comments are duplicated x4, then others follow; check the HTML, you can replace comments in app with this for testing:
[comments.txt](https://github.com/biosistemika/scinote-web/files/495298/comments.txt)


Step comments:
- date is missing for new comments - the problem is that the new comments are interpreted as if they occured on the same date as the previous last comment (I guess the same issue would occur for project and task comments, if they appeared)
screenshot from 2016-09-27 3
- date is displayed before each 10-th comment, i.e. when you press "Show more", another date is displayed, even if it's the same date as the one before - a specific date should display only once, before the first comment on that day. I've talked with Darja, and she said that now the dates should be beside time on every comment, and not standalone as are they now.

Result comments:
- date is missing for new comments - the problem is that the new comments are interpreted as if they occured on the same date as the previous last comment (I guess the same issue would occur for project and task comments, if they appeared)

**Comment by ZmagoD** [comment]
hmm this stuff already worked before so I didn't tested I will take a look at that. It is an easy fix I think I had forgot to change the logic in the list.html

File: app/assets/javascripts/comments.js [hunk 2]
```diff
@@ -1,151 +1,319 @@
...
-  // Reposition dropdown to the left
-  // (only do this when using parent offset)
-  if (useParentOffset) {
-    $(document).on("shown.bs.dropdown", ".dropdown-comment", function() {
-      var $el = $(this);
-      var menu = $el.find(".dropdown-menu");
-      var leftPos = $el.offset().left;
-      if (leftPos + menu.width() > $(window).width()) {
-        menu.offset({ left: leftPos - menu.width() });
+var Comments = (function() {
```
Comment (mz3944) [inline app/assets/javascripts/comments.js]: I would remove _var Comments =_ and youst add comment saying what it represents.

File: app/assets/javascripts/comments.js [hunk 3]
```diff
@@ -1,151 +1,319 @@
...
+  function initCommentsLink($el) {
+
+    $el.find('.btn-more-comments')
+    .on('ajax:success', function (e, data) {
+      if (data.html) {
+        var list = $(this).parents('ul');
+        var moreBtn = list.find('.btn-more-comments');
+        var listItem = moreBtn.parents('li');
+        $(data.html).insertAfter(listItem);
+        if (data.results_number < data.per_page) {
```
Comment (mz3944) [inline app/assets/javascripts/comments.js]: I guess you should use camel case instead...

File: app/assets/javascripts/comments.js [hunk 4]
```diff
@@ -1,151 +1,319 @@
...
+          moreBtn.remove();
+        } else {
+          moreBtn.attr('href', data.more_url);
```
Comment (mz3944) [inline app/assets/javascripts/comments.js]: I guess you should use camel case instead...

File: app/assets/javascripts/comments.js [hunk 5]
```diff
@@ -1,151 +1,324 @@
...
+                    });
+                });
+            });
           }
-          commentEl.remove();
+    });
+  }
+
+  function initCommentOptions(scrollableContainer, useParentOffset) {
+    useParentOffset = (typeof useParentOffset !== 'undefined') ? useParentOffset : true;
```
Comment (mz3944) [inline app/assets/javascripts/comments.js]: Fix it. You can shorten it with __.isUndefined_.

File: app/assets/javascripts/comments.js [hunk 6]
```diff
@@ -1,151 +1,324 @@
...
+      var parent = $(scrollableContainer);
+
+      if ($target.length) {
+        scrollCommentOptions(parent.find('.dropdown-comment'), useParentOffset);
+      }
+    }, true);
+  }
+
+  function scrollCommentOptions(selector, useParentOffset) {
+    useParentOffset = (typeof useParentOffset !== 'undefined') ? useParentOffset : true;
```
Comment (mz3944) [inline app/assets/javascripts/comments.js]: Fix it. You can shorten it with __.isUndefined_.

**Comment by mz3944** [comment]
The mobile devices aren't an issue, because word wrapping was implemented recently.

**Comment by mz3944** [comment]
Note that date duplicates and no current date showing should be fixed with #PR #171.

File: app/assets/javascripts/comments.js [hunk 7]
```diff
@@ -1,151 +1,340 @@
...
+                    });
+                });
+            });
           }
-          commentEl.remove();
+    });
+  }
+
+  function initCommentOptions(scrollableContainer, useParentOffset) {
+    if ( ! _.isUndefined(typeof useParentOffset) ) {
```
Comment (ZmagoD) [inline app/assets/javascripts/comments.js]: :sob:  is nod defined

**Comment by mz3944** [comment]
For new comments date is not shown. If exactly 9 comments are left to be shown and "Show more" is clicked, the button remains on top, although no more comments are left to be shown, so it should disappear.

File: app/assets/javascripts/comments.js (hunk 7 above)
Comment (mz3944) [inline app/assets/javascripts/comments.js]: @jbargu get off of this PR! It's mine!

Comment (jbargu) [inline app/assets/javascripts/comments.js]: @mz3944 Or what? You gonna write me an angry message on email?! You gonna tell me mom? Boo-freaking-hoo! Look at me, I'm shaking. :fearful:

...*whisper* (please baby don't hurt me, don't hurt me, no more)

**Comment by ZmagoD** [comment]
@mz3944 the "Show more" button is fixed now. For the comment date I have left it like it was before.

**Comment by mz3944** [comment]
Go ahead Monica, merge!

**Pull request merged**